src/Stratis.Bitcoin/P2P/PeerSelector.cs

```diff
@@ -145,18 +153,69 @@ private IEnumerable<PeerAddress> SelectPreferredPeers()
         }
 
         /// <inheritdoc/>
-        public IEnumerable<PeerAddress> SelectPeers(int peerCount)
+        public IEnumerable<PeerAddress> SelectPeersForDiscovery(int peerCount)
         {
-            // If there are no peers, just one or if the amount peers is less than
-            // the amount of peers asked for, just return the list.
+            // If there are no peers or just one then just return the list.
             if (!this.peerAddresses.Any() || this.peerAddresses.Count == 1)
                 return this.peerAddresses.Select(pa => pa.Value);
 
-            // Randomly order the list of peers and return the amount
-            // asked for.
+            // Randomly order the list of peers and return the amount asked for.
             var allPeers = this.peerAddresses.OrderBy(p => this.random.Next());
             return allPeers.Select(p => p.Value).Take(1000);
         }
+
+        /// <inheritdoc/>
+        public IEnumerable<PeerAddress> SelectPeersForGetAddrPayload(int peerCount)
+        {
+            // If there are no peers or just one then just return the list.
+            if (!this.peerAddresses.Any() || this.peerAddresses.Count == 1)
+                return this.peerAddresses.Select(pa => pa.Value);
+
+            var peersToReturn = new List<PeerAddress>();
+
+            var connectedAndHandshaked = this.peerAddresses.Connected().Concat(this.peerAddresses.Handshaked()).OrderBy(p => this.random.Next());
+            var freshAndAttempted = this.peerAddresses.Attempted().Concat(this.peerAddresses.Fresh()).OrderBy(p => this.random.Next());
+
+            //If there are connected and/or handshaked peers in the address list,
+            //we need to split the list 50 / 50 between them and
+            //peers we have not yet connected to and/or that are fresh.
```

Comment:
Again, as per comment before, I'm not understanding why you want to return peers that we haven't already at least gone through the handshake process (and then subsequently kept 'live' by pinging).  Up until that point, all other peers are purely speculative as to whether they are any good or not.  Would you want to return a list of peers to another node containing ones that you haven't performed 'due diligence' on yourself?  Apologies if I've missed the reason for this as I see this logic has already existed before and maybe there was a good reason for it that I can't immediately fathom.

Reply:
Again, we can't JUST propagate good peers throughout the network. New peers, that are potentially good ones, also needs to be given a chance to get connected to. In this method, which is only used for GetAddrPayload, we'll "fill up" the list to be provided, first with handshaked/connected and then with attempted/fresh peers. This ensures that there's always an even (fair) distribution of peers on the network.

Reply:
I would argue that you only really want to propagate good peers :)  For peers that you know about (recently connected to you or ones you received from another peer and for which you haven't yet established whether they are good or not), you have a duty to determine how 'good' they are, which could be done in the background, independent of propagating the good peers when required to do so.  I guess propagating transactions to peers should also only be done using good peers.  However, having said that, I understand your reasons for the logic of mixing the quality of the peers in the response payload.

Reply:
My question would be what is core doing?
We probably want to do the same I think this cod is responsible for address selection and logic
https://github.com/bitcoin/bitcoin/blob/5961b23898ee7c0af2626c46d5d70e80136578d3/src/net.cpp

Reply:
This is the getaddr logic 
https://github.com/bitcoin/bitcoin/blob/5961b23898ee7c0af2626c46d5d70e80136578d3/src/net_processing.cpp